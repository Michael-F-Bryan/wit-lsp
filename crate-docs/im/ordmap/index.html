<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="An ordered map."><title>im::ordmap - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="im" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.2 (25ef9e3d8 2024-04-09)" data-channel="1.77.2" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../im/index.html">im</a><span class="version">15.1.0</span></h2></div><h2 class="location"><a href="#">Module ordmap</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li></ul></section><h2><a href="../index.html">In crate im</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../im/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">im</a>::<wbr><a class="mod" href="#">ordmap</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../src/im/ord/map.rs.html#5-2649">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>An ordered map.</p>
<p>An immutable ordered map implemented as a [B-tree] <a href="https://en.wikipedia.org/wiki/B-tree">1</a>.</p>
<p>Most operations on this type of map are O(log n). A
<a href="../hashmap/struct.HashMap.html"><code>HashMap</code></a> is usually a better choice for
performance, but the <code>OrdMap</code> has the advantage of only requiring
an <a href="https://doc.rust-lang.org/std/cmp/trait.Ord.html"><code>Ord</code></a> constraint on the key, and of being
ordered, so that keys always come out from lowest to highest,
where a <a href="../hashmap/struct.HashMap.html"><code>HashMap</code></a> has no guaranteed ordering.</p>
</div></details><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.ordmap.html" title="macro im::ordmap::ordmap">ordmap</a></div><div class="desc docblock-short">Construct a map from a sequence of key/value pairs.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ConsumingIter.html" title="struct im::ordmap::ConsumingIter">ConsumingIter</a></div><div class="desc docblock-short">A consuming iterator over an ordered set.</div></li><li><div class="item-name"><a class="struct" href="struct.DiffIter.html" title="struct im::ordmap::DiffIter">DiffIter</a></div><div class="desc docblock-short">An iterator over the differences between two maps.</div></li><li><div class="item-name"><a class="struct" href="struct.Iter.html" title="struct im::ordmap::Iter">Iter</a></div><div class="desc docblock-short">An iterator over the key/value pairs of a map.</div></li><li><div class="item-name"><a class="struct" href="struct.Keys.html" title="struct im::ordmap::Keys">Keys</a></div><div class="desc docblock-short">An iterator ove the keys of a map.</div></li><li><div class="item-name"><a class="struct" href="struct.NodeDiffIter.html" title="struct im::ordmap::NodeDiffIter">NodeDiffIter</a></div><div class="desc docblock-short">An iterator over the differences between two ordered sets.</div></li><li><div class="item-name"><a class="struct" href="struct.OccupiedEntry.html" title="struct im::ordmap::OccupiedEntry">OccupiedEntry</a></div><div class="desc docblock-short">An entry for a mapping that already exists in the map.</div></li><li><div class="item-name"><a class="struct" href="struct.OrdMap.html" title="struct im::ordmap::OrdMap">OrdMap</a></div><div class="desc docblock-short">An ordered map.</div></li><li><div class="item-name"><a class="struct" href="struct.OrdMapPool.html" title="struct im::ordmap::OrdMapPool">OrdMapPool</a></div><div class="desc docblock-short">A memory pool for the appropriate node type.</div></li><li><div class="item-name"><a class="struct" href="struct.RangedIter.html" title="struct im::ordmap::RangedIter">RangedIter</a></div><div class="desc docblock-short">An iterator over an ordered set.</div></li><li><div class="item-name"><a class="struct" href="struct.VacantEntry.html" title="struct im::ordmap::VacantEntry">VacantEntry</a></div><div class="desc docblock-short">An entry for a mapping that does not already exist in the map.</div></li><li><div class="item-name"><a class="struct" href="struct.Values.html" title="struct im::ordmap::Values">Values</a></div><div class="desc docblock-short">An iterator over the values of a map.</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.DiffItem.html" title="enum im::ordmap::DiffItem">DiffItem</a></div><div class="desc docblock-short">A description of a difference between two ordered maps.</div></li><li><div class="item-name"><a class="enum" href="enum.Entry.html" title="enum im::ordmap::Entry">Entry</a></div><div class="desc docblock-short">A handle for a key and its associated value.</div></li><li><div class="item-name"><a class="enum" href="enum.NodeDiffItem.html" title="enum im::ordmap::NodeDiffItem">NodeDiffItem</a></div><div class="desc docblock-short">A description of a difference between two ordered sets.</div></li></ul></section></div></main></body></html>