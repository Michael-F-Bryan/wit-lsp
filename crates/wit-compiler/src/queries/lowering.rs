use im::{OrdMap, Vector};

use crate::{
    ast::{self, AstNode, HasAttr, HasIdent},
    diagnostics::{Diagnostic, Diagnostics, Location},
    hir,
    pointer::{
        EnumIndex, FlagsIndex, FuncItemIndex, GetByIndex, InterfaceIndex, Pointer, RecordIndex,
        ResourceIndex, ScopeIndex, TypeAliasIndex, VariantIndex, WorldIndex,
    },
    queries::{SourceFile, Workspace},
    Db, Text, Tree,
};

/// Parse a file and lower it from its [`crate::ast`] representation to the
/// [`crate::hir`] representation.
///
/// Nodes that contain syntactic or semantic errors will be ignored and a
/// corresponding [`Diagnostic`] will be emitted.
#[salsa::tracked]
#[tracing::instrument(level = "debug", skip_all, fields(file = %file.path(db)))]
pub fn lower(db: &dyn Db, _ws: Workspace, file: SourceFile) -> hir::Package {
    let ast = crate::queries::parse(db, file);
    let root = ast.source_file(db);
    let src = ast.src(db);

    let items = crate::queries::file_items(db, file);

    let mut worlds = OrdMap::new();
    let mut interfaces = OrdMap::new();

    for index in items.worlds_by_name(db).values().copied() {
        if let Some(world) = lower_world(db, file, index) {
            worlds.insert(index, world);
        }
    }

    for index in items.interfaces_by_name(db).values().copied() {
        if let Some(interface) = lower_interface(db, file, index) {
            interfaces.insert(index, interface);
        }
    }

    hir::Package {
        decl: root.package_opt().and_then(|d| lower_package_decl(d, src)),
        interfaces,
        worlds,
    }
}

fn lower_package_decl(node: ast::PackageDecl, src: &str) -> Option<hir::PackageDeclaration> {
    let docs = node.docs(src);

    let package_name = node.fully_qualified_package_name()?;
    let package = package_name.package()?.identifier()?.value(src).into();
    let path = package_name
        .path()?
        .iter_identifiers()
        .map(|p| p.value(src).into())
        .collect();
    let version = package_name.version_opt().map(|s| s.value(src).into());

    Some(hir::PackageDeclaration {
        docs,
        package,
        path,
        version,
    })
}

#[salsa::tracked]
pub(crate) fn lower_world(db: &dyn Db, file: SourceFile, index: WorldIndex) -> Option<hir::World> {
    let ast = crate::queries::parse(db, file);
    let tree = ast.tree(db);
    let items = crate::queries::file_items(db, file);

    let meta = items.get_by_index(db, index);
    let _node = meta.location(db).lookup(tree);
    let _item_definitions = meta.items(db);

    todo!();
}

#[salsa::tracked]
pub(crate) fn lower_interface(
    db: &dyn Db,
    file: SourceFile,
    index: InterfaceIndex,
) -> Option<hir::Interface> {
    let ast = crate::queries::parse(db, file);
    let tree = ast.tree(db);
    let items = crate::queries::file_items(db, file);

    let meta = items.get_by_index(db, index);
    let node = meta.location(db).lookup(tree);
    let item_definitions = meta.items(db);

    let mut items = Vector::new();
    let scope = ScopeIndex::Interface(index);

    for ix in item_definitions.iter_enums() {
        if let Some(item) = lower_enum(db, file, scope, ix) {
            items.push_back(item.into());
        }
    }

    for ix in item_definitions.iter_flags() {
        if let Some(item) = lower_flags(db, file, scope, ix) {
            items.push_back(item.into());
        }
    }

    for ix in item_definitions.iter_functions() {
        if let Some(item) = lower_func_item(db, file, index, ix) {
            items.push_back(item.into());
        }
    }

    for ix in item_definitions.iter_records() {
        if let Some(item) = lower_record(db, file, scope, ix) {
            items.push_back(item.into());
        }
    }

    for ix in item_definitions.iter_resources() {
        if let Some(item) = lower_resource(db, file, scope, ix) {
            items.push_back(item.into());
        }
    }

    for ix in item_definitions.iter_typedefs() {
        if let Some(item) = lower_type_alias(db, file, scope, ix) {
            items.push_back(item.into());
        }
    }

    for ix in item_definitions.iter_variants() {
        if let Some(item) = lower_variant(db, file, scope, ix) {
            items.push_back(item.into());
        }
    }

    Some(hir::Interface {
        name: meta.name(db),
        docs: node.docs(ast.src(db)),
        items,
    })
}

#[salsa::tracked]
pub(crate) fn lower_enum(
    db: &dyn Db,
    file: SourceFile,
    scope: ScopeIndex,
    index: EnumIndex,
) -> Option<hir::Enum> {
    let ast = crate::queries::parse(db, file);
    let ptr = match scope {
        ScopeIndex::Interface(interface) => file.get_by_index(db, (interface, index)),
        ScopeIndex::World(world) => file.get_by_index(db, (world, index)),
    };
    let src = ast.src(db);

    let tree = ast.tree(db);
    let node = ptr.lookup(tree);
    let name = node.identifier(src)?.into();

    let cases = node
        .iter_cases()
        .filter_map(|case| lower_enum_case(case, src))
        .collect();

    Some(hir::Enum {
        name,
        docs: node.docs(src),
        index,
        cases,
    })
}

fn lower_enum_case(case: ast::EnumCase<'_>, src: &str) -> Option<hir::EnumCase> {
    Some(hir::EnumCase {
        docs: case.docs(src),
        name: case.identifier(src)?.into(),
    })
}

#[salsa::tracked]
pub(crate) fn lower_flags(
    db: &dyn Db,
    file: SourceFile,
    scope: ScopeIndex,
    index: FlagsIndex,
) -> Option<hir::Flags> {
    let ast = crate::queries::parse(db, file);
    let ptr = match scope {
        ScopeIndex::Interface(interface) => file.get_by_index(db, (interface, index)),
        ScopeIndex::World(world) => file.get_by_index(db, (world, index)),
    };
    let src = ast.src(db);

    let tree = ast.tree(db);
    let node = ptr.lookup(tree);
    let name = node.identifier(src)?.into();

    let cases = node
        .iter_cases()
        .filter_map(|case| lower_flags_case(case, src))
        .collect();

    Some(hir::Flags {
        name,
        docs: node.docs(src),
        index,
        cases,
    })
}

fn lower_flags_case(case: ast::FlagsCase<'_>, src: &str) -> Option<hir::FlagsCase> {
    Some(hir::FlagsCase {
        docs: case.docs(src),
        name: case.identifier(src)?.into(),
    })
}

#[salsa::tracked]
pub(crate) fn lower_func_item(
    db: &dyn Db,
    file: SourceFile,
    interface: InterfaceIndex,
    index: FuncItemIndex,
) -> Option<hir::FuncItem> {
    let ast = crate::queries::parse(db, file);
    let tree = ast.tree(db);
    let node = file.get_by_index(db, (interface, index)).lookup(tree);
    let src = ast.src(db);

    let name = node.identifier(src)?.into();
    let func_type = node.ty()?;

    let mut params = Vector::new();

    for param in func_type.params()?.iter_params() {
        let param = lower_param(db, src, interface.into(), tree, param)?;
        params.push_back(param);
    }

    let return_value = func_type
        .result_opt()
        .and_then(|r| lower_return_value(db, src, file, interface.into(), tree, r));

    Some(hir::FuncItem {
        index,
        name,
        docs: node.docs(src),
        params,
        return_value,
    })
}

fn lower_param(
    db: &dyn Db,
    src: &str,
    scope: ScopeIndex,
    tree: &Tree,
    param: ast::NamedType<'_>,
) -> Option<hir::Parameter> {
    let name = param.identifier(src)?.into();
    let ty = param.ty()?;
    let ty = resolve_type(db, src, scope, tree, ty)?;

    Some(hir::Parameter { name, ty })
}

fn lower_return_value(
    db: &dyn Db,
    src: &str,
    file: SourceFile,
    scope: ScopeIndex,
    tree: &Tree,
    ret: ast::ResultList<'_>,
) -> Option<hir::ReturnValue> {
    if let Some(ty) = ret.ty_opt() {
        let ty = resolve_type(db, src, scope, tree, ty)?;
        Some(hir::ReturnValue::Single(ty))
    } else if let Some(list) = ret.named_result_list_opt() {
        let mut return_types = OrdMap::new();

        for pair in list.iter_named_types() {
            let name = Text::from(pair.identifier(src)?);
            let ty_node = pair.ty()?;
            let ty = resolve_type(db, src, scope, tree, ty_node).unwrap_or(hir::Type::Error);

            match return_types.entry(name) {
                im::ordmap::Entry::Vacant(entry) => {
                    entry.insert((ty, ty_node.range()));
                }
                im::ordmap::Entry::Occupied(entry) => {
                    let path = file.path(db);
                    let location = Location::new(path, pair.range());
                    let original_definition = Location::new(path, entry.get().1);

                    let diag = Diagnostic::duplicate_name(
                        entry.key().clone(),
                        location,
                        original_definition,
                    );
                    Diagnostics::push(db, diag);
                }
            }
        }

        let return_types = return_types.into_iter().map(|(k, (v, _))| (k, v)).collect();
        Some(hir::ReturnValue::Named(return_types))
    } else {
        None
    }
}

#[salsa::tracked]
pub(crate) fn lower_record(
    _db: &dyn Db,
    _file: SourceFile,
    _scope: ScopeIndex,
    _ix: RecordIndex,
) -> Option<hir::Record> {
    todo!();
}

#[salsa::tracked]
pub(crate) fn lower_resource(
    _db: &dyn Db,
    _file: SourceFile,
    _scope: ScopeIndex,
    _ix: ResourceIndex,
) -> Option<hir::Resource> {
    todo!();
}

#[salsa::tracked]
pub(crate) fn lower_type_alias(
    db: &dyn Db,
    file: SourceFile,
    scope: ScopeIndex,
    index: TypeAliasIndex,
) -> Option<hir::TypeAlias> {
    let ast = crate::queries::parse(db, file);
    let ptr = match scope {
        ScopeIndex::Interface(interface) => file.get_by_index(db, (interface, index)),
        ScopeIndex::World(world) => file.get_by_index(db, (world, index)),
    };
    let src = ast.src(db);

    let tree = ast.tree(db);
    let node = ptr.lookup(tree);
    let name = node.identifier(src)?.into();

    let ty = node.ty()?;
    let ty = resolve_type(db, src, scope, tree, ty)?;

    Some(hir::TypeAlias {
        name,
        docs: node.docs(src),
        index,
        ty,
    })
}

#[salsa::tracked]
pub(crate) fn lower_variant(
    _db: &dyn Db,
    _file: SourceFile,
    _scope: ScopeIndex,
    _ix: VariantIndex,
) -> Option<hir::Variant> {
    todo!();
}

fn resolve_type(
    _db: &dyn Db,
    src: &str,
    _scope: ScopeIndex,
    _tree: &Tree,
    ty: ast::Ty<'_>,
) -> Option<hir::Type> {
    let resolver = TypeResolver { src };
    resolver.resolve_type(ty)
}

struct TypeResolver<'a> {
    src: &'a str,
}

impl<'a> TypeResolver<'a> {
    fn resolve_type(&self, ty: ast::Ty<'_>) -> Option<hir::Type> {
        if let Some(builtin) = ty.builtins() {
            self.resolve_builtin(builtin)
        } else if let Some(handle) = ty.handle() {
            self.resolve_handle(handle)
        } else if let Some(list) = ty.list() {
            self.resolve_list(list)
        } else if let Some(option) = ty.option() {
            self.resolve_option(option)
        } else if let Some(result) = ty.result() {
            self.resolve_result(result)
        } else if let Some(tuple) = ty.tuple() {
            self.resolve_tuple(tuple)
        } else if let Some(user_defined_type) = ty.user_defined_type() {
            self.resolve_user_defined_type(user_defined_type)
        } else {
            None
        }
    }

    fn resolve_option(&self, option: ast::Option_<'_>) -> Option<hir::Type> {
        let element_type = option.ty()?;
        let element_type = self.resolve_type(element_type)?;
        Some(hir::Type::Option(Box::new(element_type)))
    }

    fn resolve_result(&self, result: ast::Result_<'_>) -> Option<hir::Type> {
        let ok = result.ok_opt().and_then(|ty| self.resolve_type(ty));
        let err = result.err_opt().and_then(|ty| self.resolve_type(ty));

        Some(hir::Type::Result {
            ok: ok.map(Box::new),
            err: err.map(Box::new),
        })
    }

    fn resolve_tuple(&self, tuple: ast::Tuple<'_>) -> Option<hir::Type> {
        let mut types = Vector::new();

        for ty in tuple.iter_tys() {
            let ty = self.resolve_type(ty).unwrap_or(hir::Type::Error);
            types.push_back(ty);
        }

        Some(hir::Type::Tuple(types))
    }

    fn resolve_user_defined_type(
        &self,
        _user_defined_type: ast::UserDefinedType<'_>,
    ) -> Option<hir::Type> {
        todo!()
    }

    fn resolve_list(&self, list: ast::List<'_>) -> Option<hir::Type> {
        let element_type = list.ty()?;
        let element_type = self.resolve_type(element_type)?;
        Some(hir::Type::List(Box::new(element_type)))
    }

    fn resolve_handle(&self, _handle: ast::Handle<'_>) -> Option<hir::Type> {
        todo!()
    }

    fn resolve_builtin(&self, builtin: ast::Builtins<'_>) -> Option<hir::Type> {
        let name = builtin.value(self.src);

        match name {
            "u8" => Some(hir::Type::Builtin(hir::Builtin::U8)),
            "u16" => Some(hir::Type::Builtin(hir::Builtin::U16)),
            "u32" => Some(hir::Type::Builtin(hir::Builtin::U32)),
            "u64" => Some(hir::Type::Builtin(hir::Builtin::U64)),
            "s8" => Some(hir::Type::Builtin(hir::Builtin::I8)),
            "s16" => Some(hir::Type::Builtin(hir::Builtin::I16)),
            "s32" => Some(hir::Type::Builtin(hir::Builtin::I32)),
            "s64" => Some(hir::Type::Builtin(hir::Builtin::I64)),
            "float32" => Some(hir::Type::Builtin(hir::Builtin::Float32)),
            "float64" => Some(hir::Type::Builtin(hir::Builtin::Float64)),
            "char" => Some(hir::Type::Builtin(hir::Builtin::Char)),
            "bool" => Some(hir::Type::Builtin(hir::Builtin::Boolean)),
            "string" => Some(hir::Type::Builtin(hir::Builtin::String)),
            other => {
                unreachable!(
                    "Unknown builtin type, \"{other}\" at {}",
                    builtin.syntax().to_sexp()
                )
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{diagnostics::Diagnostics, Compiler};

    use super::*;

    macro_rules! lowering_tests {
        (
            $(
                $( #[$meta:meta] )*
                $name:ident : $contents:literal
            ),* $(,)?
        ) => {
            $(
                #[test]
                #[allow(unused_mut, unused_variables)]
                $( #[$meta] )*
                fn $name() {
                    let db = Compiler::default();

                    let file = SourceFile::new(
                        &db,
                        format!("{}.wit", stringify!($name)).into(),
                        $contents.into(),
                    );
                    let ws = Workspace::new(&db, [(file.path(&db), file)].into_iter().collect());

                    let ast = crate::queries::parse(&db, file);
                    eprintln!("{}", ast.root_node(&db).to_sexp());

                    let got = super::lower(&db, ws, file);
                    let diags = super::lower::accumulated::<Diagnostics>(&db, ws, file);

                    assert!(diags.is_empty(), "{diags:#?}");

                    let mut settings = insta::Settings::clone_current();

                    #[derive(serde::Serialize)]
                    struct Info<'a> {
                        src: &'a str,
                        ast: String,
                    }
                    settings.set_info(&Info {
                        src: $contents,
                        ast: ast.root_node(&db).to_sexp(),
                    });
                    settings.set_omit_expression(true);

                    settings.bind(|| insta::assert_debug_snapshot!(got));
                }
            )*
        };
    }

    macro_rules! lowering_error_tests {
        (
            $(
                $( #[$meta:meta] )*
                $name:ident : $contents:literal
            ),* $(,)?
        ) => {
            $(
                #[test]
                #[allow(unused_mut, unused_variables)]
                $( #[$meta] )*
                fn $name() {
                    let db = Compiler::default();

                    let file = SourceFile::new(
                        &db,
                        format!("{}.wit", stringify!($name)).into(),
                        $contents.into(),
                    );
                    let ws = Workspace::new(&db, [(file.path(&db), file)].into_iter().collect());

                    let ast = crate::queries::parse(&db, file);
                    let diags = super::lower::accumulated::<Diagnostics>(&db, ws, file);

                    assert_ne!(diags.len(), 0, "No diagnostics emitted");

                    #[derive(serde::Serialize)]
                    struct Info<'a> {
                        src: &'a str,
                        ast: String,
                    }

                    let mut settings = insta::Settings::clone_current();
                    settings.set_info(&Info {
                        src: $contents,
                        ast: ast.root_node(&db).to_sexp(),
                    });
                    settings.set_omit_expression(true);

                    settings.bind(|| insta::assert_debug_snapshot!(diags));
                }
            )*
        };
    }

    lowering_tests! {
        lower_an_empty_file: "",
        lower_package_with_docs: "/// This is a package.\npackage wasi:filesystem@1.2.3;",
        empty_interface: "interface empty {}" ,
        func_with_no_arguments: "interface i { f: func(); }",
        func_with_one_argument: "interface i { f: func(message: string); }",
        func_with_multiple_arguments: "interface i { f: func(first: u32, second: string, third: list<bool>); }",
        func_with_return_value: "interface i { f: func() -> bool; }",
        func_with_named_return_values: "interface i { f: func() -> (a: u32, b: option<string>); }",
        interface_with_builtin_type: "interface i { type x = u32; }",
        empty_enum: "interface i { enum empty {} }",
        enum_with_one_element: "interface i { enum foo { first } }",
        enum_with_multiple_elements: "interface i {
            /// This is an enum.
            enum foo {
                /// first case.
                first,
                /// Second case.
                /// and another
                /// doc-comment.
                second,
             }
         }",
        empty_flags: "interface i { flags empty {} }",
        flags_with_one_element: "interface i { flags foo { first } }",
        flags_with_multiple_elements: "interface i {
            /// This is a flags.
            flags foo {
                /// first case.
                first,
                /// Second case.
                /// and another
                /// doc-comment.
                second,
             }
         }",
        empty_tuple: "interface i { type x = tuple<>; }",
        tuple_with_single_element: "interface i { type x = tuple<string>; }",
        tuple_with_multiple_elements: "interface i { type x = tuple<string, bool, u32>; }",
        result_with_ok_and_error: "interface i { type x = result<bool, string>; }",
        result_with_empty_ok: "interface i { type x = result<_>; }",
        bare_result: "interface i { type x = result; }",
        result_with_just_error: "interface i { type x = result<_, string>; }",
        list: "interface i { type x = list<u32>; }",
        option: "interface i { type x = option<u32>; }",

        #[ignore]
        empty_world: "world empty {}",
        #[ignore]
        world_with_function_export: "world console { export run: func(); }",
        #[ignore]
        world_with_external_export: "world with-import {
            export wasi:filesystem/filesystem;
        }",
        #[ignore]
        world_with_external_import: "world with-import {
            import wasi:filesystem/filesystem;
        }"
    }

    lowering_error_tests! {
        syntax_errors_are_emitted: "#$",
        #[ignore]
        refer_to_unknown_type: "interface i { type x = this-does-not-exist; }",
        #[ignore]
        recursive_records_are_not_allowed: "interface i { record recursive { inner: recursive } }",
        #[ignore]
        reference_cycles_are_not_allowed: "interface i {
            record first { second: second }
            record second { first: first }
        }",
        #[ignore]
        duplicate_identifiers: "interface i { record foo {} variant foo {} }",
    }
}
