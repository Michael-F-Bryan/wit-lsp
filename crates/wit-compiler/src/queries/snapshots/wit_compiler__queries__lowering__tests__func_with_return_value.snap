---
source: crates/wit-compiler/src/queries/lowering.rs
info:
  src: "interface i { f: func() -> bool; }"
  ast: "(source_file (top_level_item (interface_item name: (identifier) items: (interface_items (func_item name: (identifier) ty: (func_type params: (param_list) result: (result_list (ty (builtins)))))))))"
---
Package {
    decl: None,
    worlds: {},
    interfaces: {
        InterfaceIndex(0): Interface {
            name: i,
            docs: None,
            items: [
                Func(
                    FuncItem {
                        name: f,
                        index: FuncItemIndex(0),
                        docs: None,
                        params: [],
                        return_value: Some(
                            Single(
                                Builtin(
                                    Boolean,
                                ),
                            ),
                        ),
                    },
                ),
            ],
        },
    },
}
